### PR TITLE
Feature nullness - support overrides of nullness annotation in the imported object hierarchy

### PR DESCRIPTION
**Pull request by T-Gro**
In a nutshell:
### Problem 1
Object.ToString() -> string?
Boolean.ToString() -> string  (no null).

And many others, e.g. StringBuilder, Exception, etc.

This lead to a changed strategy in method resolution, preferring overrides instead of base methods.
This however introduced a problem in the case of "partial property override", i.e. an abstract get+set, but override only one of them.
Therefore the logic for property equivalence was changed to differentiate the hasgetter/hassetter across properties.
And method resolution 'betterness' enhanced to prefer the derived implementation in case of conflicts introduced that way.
(otherwise "no unique overload" was being shown if type hieararchy get e.g. get+set  |  get | set along the way)

### Problem 2
Nullness warning was not emited when accessing members of T|null values if those members stemmed from a supertype.
E.g. DirectoryInfo.Name was the case.
This was implicitely happening because most BCL classes have [Nullable(0)] on them, which was applied on the supertype causing it to be ambivalent.

And our prior algebra for combining nullnesses did not change ambivalent to be `WithNull` when combined with it.

### CODE REVIEW

File: src/Compiler/Checking/infos.fs [hunk 1]
```diff
@@ -2429,6 +2429,8 @@ let MethInfosEquivByNameAndPartialSig erasureFlag ignoreFinal g amap m (minfo: M
 
 /// Used to hide/filter members from base classes based on signature
 let PropInfosEquivByNameAndPartialSig erasureFlag g amap m (pinfo: PropInfo) (pinfo2: PropInfo) =
+    //(pinfo.HasGetter = pinfo2.HasGetter) &&
+    //(pinfo.HasSetter = pinfo2.HasSetter) &&
     pinfo.PropertyName = pinfo2.PropertyName &&
```
Comment (T-Gro) [inline src/Compiler/Checking/infos.fs:2434]:
```suggestion
    pinfo.PropertyName = pinfo2.PropertyName &&
```

File: src/Compiler/Checking/infos.fs [hunk 2]
```diff
@@ -2440,12 +2442,12 @@ let MethInfosEquivByNameAndSig erasureFlag ignoreFinal g amap m minfo minfo2 =
     let (CompiledSig(_, retTy, formalMethTypars, _)) = CompiledSigOfMeth g amap m minfo
     let (CompiledSig(_, retTy2, formalMethTypars2, _)) = CompiledSigOfMeth g amap m minfo2
     match retTy, retTy2 with
-    | None, None -> true
+    | None, None -> true
```
Comment (T-Gro) [inline src/Compiler/Checking/infos.fs:2445]:
```suggestion
    | None, None -> true
```

File: src/Compiler/Checking/infos.fs [hunk 3]
```diff
@@ -2440,12 +2442,12 @@ let MethInfosEquivByNameAndSig erasureFlag ignoreFinal g amap m minfo minfo2 =
...
     let (CompiledSig(_, retTy2, formalMethTypars2, _)) = CompiledSigOfMeth g amap m minfo2
     match retTy, retTy2 with
-    | None, None -> true
+    | None, None -> true                            
     | Some retTy, Some retTy2 -> typeAEquivAux erasureFlag g (TypeEquivEnv.FromEquivTypars formalMethTypars formalMethTypars2) retTy retTy2
     | _ -> false
 
 /// Used to hide/filter members from super classes based on signature
-let PropInfosEquivByNameAndSig erasureFlag g amap m (pinfo: PropInfo) (pinfo2: PropInfo) =
+let PropInfosEquivByNameAndSig erasureFlag g amap m (pinfo: PropInfo) (pinfo2: PropInfo) =
```
Comment (T-Gro) [inline src/Compiler/Checking/infos.fs:2450]:
```suggestion
let PropInfosEquivByNameAndSig erasureFlag g amap m (pinfo: PropInfo) (pinfo2: PropInfo) =
```

File: src/Compiler/TypedTree/TypedTreeBasics.fs [hunk 4]
```diff
@@ -245,7 +245,11 @@ let rec stripUnitEqnsAux canShortcut unt =
 let combineNullness (nullnessOrig: Nullness) (nullnessNew: Nullness) = 
     match nullnessOrig.Evaluate() with
     | NullnessInfo.WithoutNull -> nullnessNew
-    | NullnessInfo.AmbivalentToNull -> nullnessOrig
+    | NullnessInfo.AmbivalentToNull ->
+        match nullnessNew.Evaluate() with
+        | NullnessInfo.WithoutNull -> nullnessOrig
+        | NullnessInfo.AmbivalentToNull -> nullnessOrig
+        | NullnessInfo.WithNull -> nullnessNew
```
Comment (T-Gro) [inline src/Compiler/TypedTree/TypedTreeBasics.fs:252]: The combination of orig=Ambivalent and New=WithNull is what was happening in a reported bug.

The situation was with DirectoryInfo (extends FileSystemInfo):

let d : DirectoryInfo | null = null
d.Name  // .Name is a property from the super type


No warning was raised.
The reason was, the .Name access combined the ambivalent nullness of the supertype info (as imported from IL metadata) and the WithNull nullness of the `d` value.

File: src/Compiler/TypedTree/TypedTreeOps.fs [hunk 5]
```diff
@@ -9009,6 +9024,7 @@ let intrinsicNullnessOfTyconRef g (tcref: TyconRef) =
 
 let nullnessOfTy g ty =
     ty
+    // stripping is where the evil is for abbrevs
     |> stripTyEqns g
```
Comment (T-Gro) [inline src/Compiler/TypedTree/TypedTreeOps.fs:9028]:
```suggestion
    ty
    |> stripTyEqns g
```

**Comment by isaacabraham** [comment]
Can confirm both of these appear to now work:

```fsharp
let aBool = true
let aBoolString = aBool.ToString() // string

let aBoolObj = box aBool
let aBoolObjString = aBoolObj.ToString() // string | null
```

and

```fsharp
let n : DirectoryInfo | null = DirectoryInfo ""
let o = n.Exists // FS3261
```

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**